plugins/ExtendedTools/objmgr.c

```diff
@@ -420,393 +665,1576 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+                        PhMoveReference(&entry->TargetDrvUp, driverNameUp);
+                    }
+                }
 
-    if (PhEqualString2(Context->Object->TypeName, L"ALPC Port", TRUE))
-    {
-        static PH_INITONCE initOnce = PH_INITONCE_INIT;
-        NTSTATUS (NTAPI* NtAlpcConnectPortEx_I)(
-            _Out_ PHANDLE PortHandle,
-            _In_ POBJECT_ATTRIBUTES ConnectionPortObjectAttributes,
-            _In_opt_ POBJECT_ATTRIBUTES ClientPortObjectAttributes,
-            _In_opt_ PALPC_PORT_ATTRIBUTES PortAttributes,
-            _In_ ULONG Flags,
-            _In_opt_ PSECURITY_DESCRIPTOR ServerSecurityRequirements,
-            _Inout_updates_bytes_to_opt_(*BufferLength, *BufferLength) PPORT_MESSAGE ConnectionMessage,
-            _Inout_opt_ PSIZE_T BufferLength,
-            _Inout_opt_ PALPC_MESSAGE_ATTRIBUTES OutMessageAttributes,
-            _Inout_opt_ PALPC_MESSAGE_ATTRIBUTES InMessageAttributes,
-            _In_opt_ PLARGE_INTEGER Timeout
-            ) = NULL;
-
-        if (PhBeginInitOnce(&initOnce))
-        {
-            NtAlpcConnectPortEx_I = PhGetModuleProcAddress(L"ntdll.dll", "NtAlpcConnectPortEx");
-            PhEndInitOnce(&initOnce);
-        }
+                // The device might be a PDO... Query the PnP manager for the friendly name of the device.
+                if (!entry->Target)
+                {
+                    PPH_STRING devicePdoName;
+                    PPH_STRING devicePdoName2;
 
-        if (NtAlpcConnectPortEx_I)
-        {
-            status = NtAlpcConnectPortEx_I(
+                    if (devicePdoName = PhGetPnPDeviceName(deviceName))
+                    {
+                        ULONG_PTR column_pos = PhFindLastCharInString(devicePdoName, 0, L':');
+                        devicePdoName2 = PhSubstring(devicePdoName, 0, column_pos + 1);
+                        devicePdoName2->Buffer[column_pos - 4] = L'[', devicePdoName2->Buffer[column_pos] = L']';
+                        PhMoveReference(&entry->Target, devicePdoName2);
+                        entry->PdoDevice = TRUE;
+                        PhDereferenceObject(devicePdoName);
+                    }
+                }
+
+                PhDereferenceObject(deviceName);
+            }
+            break;
+        case EtObjectAlpcPort:
+            {
+                // Using fast connect to port since we only need query connection OwnerProcessId
+                if (!NT_SUCCESS(status = EtObjectManagerOpenHandle(&objectHandle, &objectContext, READ_CONTROL, 0)))
+                {
+                    // On failure try to open real (rare)
+                    status = EtObjectManagerOpenHandle(&objectHandle, &objectContext, READ_CONTROL, OBJECT_OPENSOURCE_ALPCPORT);
+                }
+                if (NT_SUCCESS(status))
+                {
+                    KPH_ALPC_COMMUNICATION_INFORMATION connectionInfo;
+
+                    if (NT_SUCCESS(status = KphAlpcQueryInformation(
```

Comment:
Oddity here - this needs a check for `KsiLevel()` else an assert will trigger on the uninitialized free list. I want to eventually make it so the initialization of the driver isn't a requirement for trying to call the APIs, but that's not how it works today. Please add a check if `KsiLevel() >= KphLevelMed` before calling `KphAlpcQueryInformation`.

Reply:
Alpc port and driver target resolving only enabled with `BOOLEAN useKsi = KsiLevel() >= KphLevelMed;`

```
        else if (entry->EtObjectType == EtObjectDevice ||       // allow resolving without driver for PhGetPnPDeviceName()
            entry->EtObjectType == EtObjectAlpcPort && useKsi ||
            entry->EtObjectType == EtObjectMutant ||
            entry->EtObjectType == EtObjectJob ||
            entry->EtObjectType == EtObjectDriver && useKsi)
        {
            entry->TargetIsResolving = TRUE;
        }
```

In this case, resolver code will not be queued.
```
    for (ULONG i = 0; i < Context->CurrentDirectoryList->Count; i++)
    {
        ...
        if (!entry->TargetIsResolving)
            continue;
        ...
    }
```